src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -454,10 +484,38 @@ protected virtual void CompleteTransaction(bool isCommitted)
 				// do an early exit here in such case.
 				if (!IsInActiveTransaction)
 					return;
+				var isSessionProcessing = _session.IsProcessing();
 				try
 				{
 					// Allow transaction completed actions to run while others stay blocked.
 					_bypassLock.Value = true;
+					// Ensure no other session processing is still ongoing. In case of a transaction timeout, the transaction is
+					// cancelled on a new thread even for non-distributed scopes. So, the session could be doing some processing,
+					// and will not be interrupted until attempting some usage of the connection. See #3355.
+					// Thread safety of a concurrent session BeginProcess is ensured by the Wait performed by BeginProcess.
+					if (isSessionProcessing)
+					{
+						var timeOutGuard = new Stopwatch();
+						timeOutGuard.Start();
+						while (isSessionProcessing && timeOutGuard.ElapsedMilliseconds < _systemTransactionCompletionLockTimeout)
+						{
+							// Naïve yield.
+							Thread.Sleep(10);
+							isSessionProcessing = _session.IsProcessing();
+						}
+						if (isSessionProcessing)
+						{
+							// Throwing would give up attempting to close the session if need be, which may still succeed. So,
+							// just log an error.
+							_logger.Error(
```

Comment:
Maybe this should be "downgraded" to just `Warn`.

Reply:
Done, also as a way to relaunch the CI. For some reason most Teamcity builds were not triggered.